#VRML_SIM R2025a utf8
PROTO FourWheelsRobot [
  field SFVec3f translation 0 0 0
  field SFRotation rotation 0 0 1 0
  field SFFloat bodyMass 1
  field SFString carController ""
]
{
  Robot {
    translation IS translation
    rotation IS rotation
    controller IS carController

    translation 0 0 0.05
    children [
        DEF body Shape {
        appearance PBRAppearance {
            baseColor 1 0.149126 0
            roughness 1
            metalness 0
        }
        geometry Box {
            size 0.2 0.1 0.05
        }
        }
        DEF wheel1 HingeJoint {
        jointParameters HingeJointParameters {
            axis 0 1 0
            anchor 0.05 0.06 0
        }
        device [
            RotationalMotor {
            name "wheel_lf"
            }
        ]
        endPoint DEF wheel1 Solid {
            translation 0.05 0.06 0
            rotation 1 0 0 1.5708
            children [
            DEF wheel Shape {
                appearance PBRAppearance {
                baseColor 0 0.976806 0
                roughness 1
                metalness 0
                }
                geometry Cylinder {
                height 0.02
                radius 0.04
                }
            }
            ]
            boundingObject USE wheel
            physics DEF wheel_ph Physics {
            }
        }
        }
        DEF wheel2 HingeJoint {
        jointParameters HingeJointParameters {
            axis 0 1 0
            anchor 0.05 0.06 0
        }
        device [
            RotationalMotor {
            name "wheel_rf"
            }
        ]
        endPoint Solid {
            translation 0.05 -0.06 0
            rotation 1 0 0 1.5708
            children [
            USE wheel
            ]
            name "solid(1)"
            boundingObject USE wheel
            physics USE wheel_ph
        }
        }
        DEF wheel3 HingeJoint {
        jointParameters HingeJointParameters {
            axis 0 1 0
            anchor -0.05 -0.06 0
        }
        device [
            RotationalMotor {
            name "wheel_rb"
            }
        ]
        endPoint Solid {
            translation -0.05 -0.06 0
            rotation 1 0 0 1.5708
            children [
            USE wheel
            ]
            name "solid(2)"
            boundingObject USE wheel
            physics USE wheel_ph
        }
        }
        DEF wheel4 HingeJoint {
        jointParameters HingeJointParameters {
            axis 0 1 0
            anchor -0.05 0.06 0
        }
        device [
            RotationalMotor {
            name "wheel_lb"
            }
        ]
        endPoint Solid {
            translation -0.05 0.06 0
            rotation 1 0 0 1.5708
            children [
            USE wheel
            ]
            name "solid(3)"
            boundingObject USE wheel
            physics USE wheel_ph
        }
        }
        DistanceSensor {
        translation 0.1 0.025 0
        children [
            DEF eye Shape {
            appearance PBRAppearance {
                baseColor 0.0168002 0.198352 1
                roughness 1
                metalness 0
            }
            geometry Sphere {
                radius 0.01
                subdivision 2
            }
            }
        ]
        name "ds_left"
        boundingObject USE eye
        physics DEF eye_ph Physics {
        }
        }
        DistanceSensor {
        translation 0.1 -0.025 0
        children [
            USE eye
        ]
        name "ds_right"
        boundingObject USE eye
        physics USE eye_ph
        }
    ]
    boundingObject USE body
    physics Physics {
      mass IS bodyMass
    }
    controller "four_wheeled_robot"
  }
}